google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java

```diff
@@ -0,0 +1,484 @@
+    //noinspection ConstantConditions guarded by "assertNoError" above
+    QueryDocumentSnapshot doc = e.value.getDocumentChanges().get(0).getDocument();
+    assertThat(doc.get("foo")).isEqualTo("bar");
+  }
+
+  /*
+  Starting from a query with non-empty results, modifying an existing document which matches
+  the query such that it no longer matches the query should result in a REMOVED event
+   */
+  @Test
+  public void nonEmptyResults_modifiedDocument_REMOVED()
+      throws InterruptedException, TimeoutException, ExecutionException {
+    DocumentReference testDoc = randomColl.document("doc");
+    // create our "existing non-matching document"
+    testDoc.set(map("foo", "bar")).get(5, TimeUnit.SECONDS);
+
+    // run our test
+    final Query query = randomColl.whereEqualTo("foo", "bar");
+    // register the snapshot listener for the query
+    QuerySnapshotEventListener listener = new QuerySnapshotEventListener(1, 0, 0, 1);
+    List<ListenerEvent> receivedEvents = listener.receivedEvents;
+    ListenerRegistration registration = query.addSnapshotListener(listener);
+
+    try {
+      listener.eventsCDL.await();
+      testDoc.set(map("bar", "foo")).get(5, TimeUnit.SECONDS);
+      listener.eventsCDL.await(DocumentChange.Type.REMOVED);
+    } finally {
+      registration.remove();
+    }
+
+    ListenerAssertions la = listener.assertions();
+    la.noError();
+    la.eventCountIsAnyOf(Range.closed(1, 2));
+    la.addedIdsIsAnyOf(emptySet(), newHashSet("doc"));
+    la.modifiedIdsIsAnyOf(emptySet());
+    la.removedIdsIsAnyOf(emptySet(), newHashSet("doc"));
+
+    ListenerEvent e = receivedEvents.get(receivedEvents.size() - 1);
+    //noinspection ConstantConditions guarded by "assertNoError" above
+    QueryDocumentSnapshot doc = e.value.getDocumentChanges().get(0).getDocument();
+    assertThat(doc.get("foo")).isEqualTo("bar");
+  }
+
+  /**
+   * A tuple class used by {@code #queryWatch}. This class represents an event delivered to the
+   * registered query listener.
+   */
+  private static final class ListenerEvent {
+
+    @Nullable private final QuerySnapshot value;
+    @Nullable private final FirestoreException error;
+
+    ListenerEvent(@Nullable QuerySnapshot value, @Nullable FirestoreException error) {
+      this.value = value;
+      this.error = error;
+    }
+  }
+
+  private static final class EventsCDL {
+    private final CountDownLatch cdl;
+    private final EnumMap<DocumentChange.Type, CountDownLatch> cdls;
+
+    EventsCDL(int nonChangeInitial, int addedInitial, int modifiedInitial, int removedInitial) {
+      cdl = new CountDownLatch(nonChangeInitial);
+      cdls = new EnumMap<>(DocumentChange.Type.class);
+      cdls.put(DocumentChange.Type.ADDED, new CountDownLatch(addedInitial));
+      cdls.put(DocumentChange.Type.MODIFIED, new CountDownLatch(modifiedInitial));
+      cdls.put(DocumentChange.Type.REMOVED, new CountDownLatch(removedInitial));
```

Comment:
The way the Watch is used in this test (adding or modifying a document at a time) is actually deterministic. By splitting these updates across types, you add some complexity to the tests that I don't think you need.

Reply:
I tried removing the individual CountDownLatches here and ran into several tests becoming more flaky, so I'd like to leave this for now and spend more time digging into it later.